pkg/server/server.go

```diff
@@ -451,30 +449,21 @@ func NewServerWithOpts(opts ...OpenFGAServiceV1Option) (*Server, error) {
 
 	// below this point, don't throw errors or we may leak resources in tests
 
-	cycleDetectionCheckResolver := graph.NewCycleDetectionCheckResolver()
-	s.checkResolver = cycleDetectionCheckResolver
-
-	localChecker := graph.NewLocalChecker(
+	var checkBuilderOpts []graph.CheckQueryBuilderOpt
+	checkBuilderOpts = append(checkBuilderOpts, graph.WithLocalCheckerOpts(
 		graph.WithResolveNodeBreadthLimit(s.resolveNodeBreadthLimit),
-	)
-
-	cycleDetectionCheckResolver.SetDelegate(localChecker)
-	localChecker.SetDelegate(cycleDetectionCheckResolver)
+	))
 
 	if s.checkQueryCacheEnabled {
 		s.logger.Info("Check query cache is enabled and may lead to stale query results up to the configured query cache TTL",
 			zap.Duration("CheckQueryCacheTTL", s.checkQueryCacheTTL),
 			zap.Uint32("CheckQueryCacheLimit", s.checkQueryCacheLimit))
 
-		cachedCheckResolver := graph.NewCachedCheckResolver(
+		checkBuilderOpts = append(checkBuilderOpts, graph.WithCachedCheckResolverOpts(
 			graph.WithMaxCacheSize(int64(s.checkQueryCacheLimit)),
 			graph.WithLogger(s.logger),
 			graph.WithCacheTTL(s.checkQueryCacheTTL),
-		)
-		s.cachedCheckResolver = cachedCheckResolver
-
-		cachedCheckResolver.SetDelegate(localChecker)
-		cycleDetectionCheckResolver.SetDelegate(cachedCheckResolver)
+		))
```

Comment:
These bits

```go
if s.checkQueryCacheEnabled {
    checkBuilderOpts = append(checkBuilderOpts, graph.WithCachedCheckResolverOpts(
   ...
}
if s.checkDispatchThrottlingEnabled {
   checkBuilderOpts = append(checkBuilderOpts, graph.WithDispatchThrottlingCheckResolverOpts(
   ...
}
````

I think should be moved to the new code you wrote (you'll have to add new builder functions for the `enabled` flags)

We don't need to preserve the logs - on a previous commit we are now logging every single server setting on startup (https://github.com/openfga/openfga/commit/fb0ed0d53ad022e3534d1ca181dc714e6d922ed9)

Reply:
But wouldn't that mean we have to pass the server configs there. I feel that would break Single responsibility.

Reply:
In my view, today we have the responsibility split - the server and the builder both kind of know how to build the chain in the right order. My idea was that only the builder knows about the right order, which would mean it would have to accept boolean flags to know which resolvers to add to the chain.